kion/resource_project_enforcement.go

```diff
@@ -0,0 +1,326 @@
+			Severity: diag.Error,
+			Summary:  "Failed Project Enforcement Creation",
+			Detail:   fmt.Sprintf("Error: %v\nItem: %v", errors.New("received item ID of 0"), post),
+		})
+		return diags
+	}
+
+	d.SetId(strconv.Itoa(resp.RecordID))
+
+	return diags
+}
+
+func resourceProjectEnforcementRead(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	var diags diag.Diagnostics
+	client := m.(*hc.Client)
+
+	enforcementID := d.Id()
+	projectID, ok := d.GetOk("project_id")
+	if !ok {
+		return diag.Errorf("Invalid or missing project ID")
+	}
+
+	enforcementIDInt, err := strconv.Atoi(enforcementID)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	resp := new(hc.ProjectEnforcementResponse)
+	err = client.GET(fmt.Sprintf("/v3/project/%d/enforcement", projectID), resp)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	if len(resp.Data) == 0 {
+		return diag.Errorf("No Project Enforcement data found for Project ID %d", projectID)
+	}
+
+	var found bool
+	for _, item := range resp.Data {
+		if int(item.ID) == enforcementIDInt {
+			diags = append(diags, safeSet(d, "description", item.Description)...)
+			diags = append(diags, safeSet(d, "timeframe", item.Timeframe)...)
+			diags = append(diags, safeSet(d, "spend_option", item.SpendOption)...)
+			diags = append(diags, safeSet(d, "amount_type", item.AmountType)...)
+			diags = append(diags, safeSet(d, "threshold_type", item.ThresholdType)...)
+			diags = append(diags, safeSet(d, "threshold", item.Threshold)...)
+			diags = append(diags, safeSet(d, "enabled", item.Enabled)...)
+			diags = append(diags, safeSet(d, "overburn", item.Overburn)...)
+			diags = append(diags, safeSet(d, "user_group_ids", item.UserGroupIds)...)
+			found = true
+			break
+		}
+	}
+
+	if !found {
+		return diag.Errorf("Enforcement ID %d not found under Project ID %d", enforcementIDInt, projectID)
+	}
+
+	return diags
+}
+
+func resourceProjectEnforcementUpdate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	var diags diag.Diagnostics
+	client := m.(*hc.Client)
+
+	projectID, ok := d.GetOk("project_id")
+	if !ok {
+		return diag.Errorf("Invalid or missing project ID")
+	}
+
+	projectIDInt := projectID.(int)
+	enforcementID := d.Id()
+
+	// Prepare the update request
+	req := hc.ProjectEnforcementUpdate{
+		Description:   d.Get("description").(string),
+		Timeframe:     d.Get("timeframe").(string),
+		SpendOption:   d.Get("spend_option").(string),
+		AmountType:    d.Get("amount_type").(string),
+		ServiceID:     hc.OptionalInt(d, "service_id"),
+		ThresholdType: d.Get("threshold_type").(string),
+		Threshold:     d.Get("threshold").(int),
+		CloudRuleID:   hc.OptionalInt(d, "cloud_rule_id"),
+		Overburn:      d.Get("overburn").(bool),
+		Enabled:       d.Get("enabled").(bool),
+	}
+
+	// Send the update request
+	endpoint := fmt.Sprintf("/v3/project/%d/enforcement/%s", projectIDInt, enforcementID)
+	err := client.PATCH(endpoint, req)
```

Comment:
If I'm reading this correctly, the PATCH endpoint will be called on every terraform apply since there is no `d.HasChanges` logic.   We only want to call PATCH if the resource actually changed.

Reply:
I see what you mean, and yeah, I missed that.  So when I was testing this locally, and I had no changes locally, and Terraform said there were no changes, my current code still went out and looked to see if there were any changes to be made?

Reply:
Yeah, that's exactly what it was doing.  Lesson learned.